core/src/main/scala/org/apache/spark/executor/Executor.scala

```diff
@@ -437,7 +441,14 @@ private[spark] class Executor(
       }
     }
 
-    val message = Heartbeat(executorId, tasksMetrics.toArray, env.blockManager.blockManagerId)
+    val executorMetrics = if (isLocal) {
+      Utils.deserialize[ExecutorMetrics](Utils.serialize(this.executorMetrics))
```

Comment:
is the idea of this to just to let you debug serialization issues even while you are running in local mode?  If that is the only reason, I think before this gets merged we probably want to get rid of it, you can always test w/ "local-cluster" which will do the serialization.